### PR TITLE
android task 4 create an experiment called item sorting

### PR DESCRIPTION
**Pull request by lttsNilesh**
Create an experiment called item_sorting
a. Add 3 groups: control, test oldest, test newest
b. Add a single parameter sort_order, which is “default”, “newest_first” or “oldest_first”
c. Implementing the different sorting by checking the experiment and parameter value for the user

### CODE REVIEW

File: android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt [hunk 1]
```diff
@@ -83,10 +85,42 @@ class DefaultTaskRepository @Inject constructor(
         }
     }
 
-    override fun getTasksStream(): Flow<List<Task>> {
-        return localDataSource.observeAll().map { tasks ->
-            withContext(dispatcher) {
-                tasks.toExternal()
+    override fun getTasksStream(sortOrderValue: Int): Flow<List<Task>> {
+        Statsig.logEvent(
+            "${StatsigUtil.ITEM_SORT} ",
```
Comment (daniel-statsig) [inline android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt:90]: You don't need the string templating for the event name here. This is going to result in the event name not matching the others. "item_sorting " vs "item_sorting"

File: android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt [hunk 2]
```diff
@@ -83,10 +85,42 @@ class DefaultTaskRepository @Inject constructor(
         }
     }
 
-    override fun getTasksStream(): Flow<List<Task>> {
-        return localDataSource.observeAll().map { tasks ->
-            withContext(dispatcher) {
-                tasks.toExternal()
+    override fun getTasksStream(sortOrderValue: Int): Flow<List<Task>> {
```
Comment (daniel-statsig) [inline android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt:88]: Use an Enum. Something like:

```
enum class TaskSortOrder {
    None,
    Alphabetical,
    NewestFirst,
    OldestFirst
}
```

File: android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt [hunk 3]
```diff
@@ -83,10 +85,42 @@ class DefaultTaskRepository @Inject constructor(
...
+        Statsig.logEvent(
+            "${StatsigUtil.ITEM_SORT} ",
+            "${StatsigUtil.SORT_ORDER} $sortOrderValue"
+        )
+        return when (sortOrderValue) {
+            1 -> {
+                localDataSource.orderAlphabetically().map { tasks ->
+                    withContext(dispatcher) {
```
Comment (daniel-statsig) [inline android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt:96]: The mapping function here is the same for all these cases. Just get the data and then map it.

```
val data = when (order) {
 TaskSortOrder.Alphabetical -> localDataSource.orderAlphabetically()
 TaskSortOrder.NewestFirst -> localDataSource.orderNewestDateWise()
 ...
} 

return data.map { tasks ->
    withContext(dispatcher) {
        tasks.toExternal()
    }
}
```

File: android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/ModelMappingExt.kt [hunk 4]
```diff
@@ -90,3 +102,9 @@ fun NetworkTask.toExternal() = toLocal().toExternal()
 
 @JvmName("networkToExternal")
 fun List<NetworkTask>.toExternal() = map(NetworkTask::toExternal)
+
+fun String.getCurrentDateTime(): String {
```
Comment (daniel-statsig) [inline android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/ModelMappingExt.kt:106]: Does this need to live on String? Can it not just be a static function

**Review by daniel-statsig — APPROVED** [review]
Approving to unblock. Please address feeback

**Review by tore-statsig — COMMENTED** [review]
Looks pretty good - I think all the updates per suggestions were made here - are you just going to land this pr and abandon the others?

Note daniels feedback

File: android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt (hunk 1 above)
Comment (lttsNilesh) [reply android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt:90]: I will remove it.

File: android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt (hunk 2 above)
Comment (lttsNilesh) [reply android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt:88]: I will use it.

File: android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt (hunk 3 above)
Comment (lttsNilesh) [reply android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/DefaultTaskRepository.kt]: Implemented.

File: android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/ModelMappingExt.kt (hunk 4 above)
Comment (lttsNilesh) [reply android-todo-list/architecture-samples-main/app/src/main/java/com/statsig/todoapp/data/ModelMappingExt.kt:106]: Will change it.

**Comment by lttsNilesh** [comment]
> Looks pretty good - I think all the updates per suggestions were made here - are you just going to land this pr and abandon the others?
> 
> Note daniels feedback

Daniel's feedback points implemented. I would go ahead with this PR since it has all the changes and review points.

**Comment by lttsNilesh** [comment]
> Approving to unblock. Please address feeback

Your review po8ints have been implemented. Thank you.

**Pull request merged**